### PR TITLE
820 - Needs endpoint

### PR DESCRIPTION
**Pull request by brownbearsoftware**
**Implement GET /people/{crn}/needs**
Retrieves needs from ARNS and groups the `identifiedNeeds` into categories suitable for our frontend

### CODE REVIEW

**Comment by brownbearsoftware** [comment]
[Get needs from Assess Risks and Needs](https://trello.com/c/a9Jtyq2t/820-get-needs-from-assess-risks-and-needs)

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/PersonNeedsTest.kt [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+                identifiedAsNeed = true,
+                needScore = 5,
+                riskOfReoffending = false
+              )
+            )
+          )
+        )
+      )
+  }
+}
```
Comment (edavey) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/PersonNeedsTest.kt:215]: this is really easy to follow. I wonder if this last test could be a little easier to parse if you used the `needScore` as hook to link the fixtures with the assertions? e.g.

```
 wiremockServer.stubFor(
      get(WireMock.urlEqualTo("/needs/crn/CRN"))
        .willReturn(
          aResponse()
            .withHeader("Content-Type", "application/json")
            .withStatus(200)
            .withBody(
              objectMapper.writeValueAsString(
                NeedsFactory()
                  .withAssessedOn(LocalDateTime.now())
                  .withIdentifiedNeeds(
                    listOf(
                      NeedFactory()
                        .withSeverity(NeedSeverity.SEVERE)
                        .withSection("SEVERE_HARM")
                        .withName("SEVERE_HARM")
                        .withOverThreshold(true)
                        .withRiskOfHarm(true)
                        .withRiskOfReoffending(true)
                        .withFlaggedAsNeed(true)
                        .withIdentifiedAsNeed(true)
                        .withNeedScore(10)
                        .produce(),
                      NeedFactory()
                        .withSeverity(NeedSeverity.STANDARD)
                        .withSection("STANDARD_HARM")
                        .withName("STANDARD_HARM")
                        .withOverThreshold(true)
                        .withRiskOfHarm(true)
                        .withRiskOfReoffending(false)
                        .withFlaggedAsNeed(true)
                        .withIdentifiedAsNeed(true)
                        .withNeedScore(2)
                        .produce(),
                      NeedFactory()
                        .withSeverity(NeedSeverity.STANDARD)
                        .withSection("RISK_OF_REOFFEND")
                        .withName("RISK_OF_REOFFEND")
                        .withOverThreshold(true)
                        .withRiskOfHarm(false)
                        .withRiskOfReoffending(true)
                        .withFlaggedAsNeed(true)
                        .withIdentifiedAsNeed(true)
                        .withNeedScore(5)
                        .produce()
                    )
                  ).produce()
              )
            )
        )
    )

    val jwt = jwtAuthHelper.createAuthorizationCodeJwt(
      subject = "username",
      authSource = "delius",
      roles = listOf("ROLE_PROBATION")
    )

    webTestClient.get()
      .uri("/people/CRN/needs")
      .header("Authorization", "Bearer $jwt")
      .exchange()
      .expectStatus()
      .isOk
      .expectBody()
      .json(
        objectMapper.writeValueAsString(
          PersonNeeds(
            linkedToRiskOfSeriousHarm = listOf(
              PersonNeed(
                section = "SEVERE_HARM",
                name = "SEVERE_HARM",
                overThreshold = true,
                riskOfHarm = true,
                flaggedAsNeed = true,
                severity = "SEVERE",
                identifiedAsNeed = true,
                needScore = 10,
                riskOfReoffending = true
              )
            ),
            linkedToReoffending = listOf(
              PersonNeed(
                section = "RISK_OF_REOFFEND",
                name = "RISK_OF_REOFFEND",
                overThreshold = true,
                riskOfHarm = false,
                flaggedAsNeed = true,
                severity = "STANDARD",
                identifiedAsNeed = true,
                needScore = 5,
                riskOfReoffending = true
              )
            ),
            notLinkedToSeriousHarmOrReoffending = listOf(
              PersonNeed(
                section = "STANDARD_HARM",
                name = "STANDARD_HARM",
                overThreshold = true,
                riskOfHarm = true,
                flaggedAsNeed = true,
                severity = "STANDARD",
                identifiedAsNeed = true,
                needScore = 2,
                riskOfReoffending = false
              )
            )
          )
        )
```

**Review by edavey — APPROVED** [review]
Very clear work, thanks.

I've made a tiny suggestion which might help readability in the complex response assertions.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/PersonNeedsTest.kt (hunk 1 above)
Comment (brownbearsoftware) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/PersonNeedsTest.kt:215]: 👍

**Pull request merged**